### ISSUE TITLE
Data.Fin and with statement

### ISSUE THREAD

**Issue by shinji-kono** [agda/agda-stdlib#2106]
I'm using Agda version 2.6.4 and standard-library-2.0

In the fixing a working code in Agda 2.6.3 and standard-library,  (simialar one with `Data.Fin.Permutation.remove`)
```agda
   p← : Fin n → Fin n 
   p← x with perm ⟨$⟩ˡ (suc x) in eq
   p← x | zero  = ⊥-elim ( sh2 perm p0=0 {x} eq )
   p← x | suc t = t

   p01 : (x : Fin n) → ?
   p01 x with perm ⟨$⟩ˡ (suc x) 
   ... | t = ?

   bad1 : (x : Fin n) → Fin (suc n)
   bad1 x = perm ⟨$⟩ˡ (suc x) 

   bad : (x : Fin n) → p← x ≡ ?
   bad with bad1 x
   ... | t = ?
```

This gives me ...

```
/Users/kono/src/galois/src/pm.agda:60,4-15
Function.Bundles.Inverse.from perm (suc x) != w of type Fin (suc n)
when checking that the type
{n : ℕ} (perm : Permutation (suc n) (suc n)) (x : Fin n)
(w : Fin (suc n))
(p0=0 : Function.Bundles.Inverse.from perm zero ≡ zero) →
(p← perm p0=0 x | w | refl) ≡
?2 (perm = perm) (p0=0 = p0=0) (x = x)
of the generated with function is well-formed
```

[pm.agda](https://gist.github.com/shinji-kono/69afbae6a24112f7fb086581979cf36b)

**Comment by andreasabel** [comment]
I think this isn't a bug in Agda, maybe some inconvenience with the new `with...in...`. I moved it here because it feels like it belongs into the _deprecation of inspect_ topic.
Also, I think here is more expertise how to prove stuff about functions defined with `with...in...`.

Here is the code from the OP with polished import statements:
```agda
{-# OPTIONS --cubical-compatible --safe #-}
module _ where

open import Data.Fin                              using (Fin; zero; suc; #_)
open import Data.Fin.Permutation                  using (Permutation; _⟨$⟩ˡ_; _⟨$⟩ʳ_; inverseˡ; inverseʳ; permutation)
open import Data.Nat.Base                         using (ℕ; suc; zero; s≤s ; z≤n )
open import Relation.Binary.PropositionalEquality using (_≡_; cong; refl; sym; module ≡-Reasoning)

open import Relation.Nullary                      using (¬_)
open import Data.Empty                            using (⊥-elim)

shlem→ : {n : ℕ} → (perm : Permutation (suc n) (suc n) ) → (p0=0 : perm ⟨$⟩ˡ (# 0) ≡ # 0 ) → (x : Fin (suc n) ) →  perm ⟨$⟩ˡ x ≡ zero → x ≡ zero
shlem→ perm p0=0 x px=0 = begin
          x                                  ≡⟨ sym ( inverseʳ perm ) ⟩
          perm ⟨$⟩ʳ ( perm ⟨$⟩ˡ x)           ≡⟨ cong (λ k →  perm ⟨$⟩ʳ k ) px=0 ⟩
          perm ⟨$⟩ʳ zero                     ≡⟨ cong (λ k →  perm ⟨$⟩ʳ k ) (sym p0=0) ⟩
          perm ⟨$⟩ʳ ( perm ⟨$⟩ˡ zero)        ≡⟨ inverseʳ perm  ⟩
          zero
       ∎  where open ≡-Reasoning

shlem← : {n : ℕ} → (perm : Permutation (suc n) (suc n) ) → (p0=0 : perm ⟨$⟩ˡ (# 0) ≡ # 0 ) → (x : Fin (suc n)) → perm ⟨$⟩ʳ x ≡ zero → x ≡ zero
shlem← perm p0=0 x px=0 =  begin
          x                                  ≡⟨ sym (inverseˡ perm ) ⟩
          perm ⟨$⟩ˡ ( perm ⟨$⟩ʳ x )          ≡⟨ cong (λ k →  perm ⟨$⟩ˡ k ) px=0 ⟩
          perm ⟨$⟩ˡ zero                     ≡⟨ p0=0  ⟩
          zero
       ∎  where open ≡-Reasoning

sh2 : {n : ℕ} → (perm : Permutation (suc n) (suc n) ) → (p0=0 : perm ⟨$⟩ˡ (# 0) ≡ # 0 ) → {x : Fin n} → ¬ perm ⟨$⟩ˡ (suc x) ≡ zero
sh2 perm p0=0 {x} eq with shlem→ perm p0=0 (suc x) eq
sh2 perm p0=0 {x} eq | ()

sh1 : {n : ℕ} → (perm : Permutation (suc n) (suc n) ) → (p0=0 : perm ⟨$⟩ˡ (# 0) ≡ # 0 ) → {x : Fin n} → ¬ perm ⟨$⟩ʳ (suc x) ≡ zero
sh1 perm p0=0 {x} eq with shlem← perm p0=0 (suc x) eq
sh1 perm p0=0 {x} eq | ()


-- 0 ∷ 1 ∷ 2 ∷ 3 ∷ [] → 0 ∷ 1 ∷ 2 ∷ []
shrink : {n : ℕ} → (perm : Permutation (suc n) (suc n) ) → perm ⟨$⟩ˡ (# 0) ≡ # 0 → Permutation n n
shrink {n} perm p0=0  = permutation p→ p← piso← piso→  where

   p→ : Fin n → Fin n
   p→ x with perm ⟨$⟩ʳ (suc x) in eq
   p→ x | zero  = ⊥-elim ( sh1 perm p0=0 {x} eq )
   p→ x | suc t = t

   p← : Fin n → Fin n
   p← x with perm ⟨$⟩ˡ (suc x) in eq
   p← x | zero  = ⊥-elim ( sh2 perm p0=0 {x} eq )
   p← x | suc t = t

   abbrv : (x : Fin n) → Fin (suc n)
   abbrv x = perm ⟨$⟩ˡ (suc x)

   --  using "with" something bound in ≡ is prohibited
   piso← : (x : Fin n ) → p→ ( p← x ) ≡ x
   piso← x = {!!}

   piso→ : (x : Fin n ) → p← ( p→ x ) ≡ x
   piso→ x = {!!}

   p01 : (x : Fin n) →  p→ ( p← x ) ≡ x
   p01 x with perm ⟨$⟩ˡ (suc x) in eq
   ... | t = {!!}
```
The task is to prove `piso←`, and `p01` is a draft that does not work because `with`-abstraction does not do the right thing here.

If I wanted to make progress on this, I probably would reformulate `p→` in terms of the inspect idiom and then see if I get the correct magic `with` incantation.

**Comment by jamesmckinna** [comment]
Ha! Thanks for this... I might also wish to loop @gallais and even notorious `with`-sceptic @omelkonian (sorry Orestis!) for moral support...

Meanwhile, my continuing tracking of @omelkonian 's issues in the main Agda issue tracker have led me to the (reluctant!) conclusion that we
* continue to eliminate uses of `inspect` from the library;
* BUT, abandon its *deprecation* for the time being...

Meanwhile, I'll scratch my head a bit harder about this example, and throw back to the main developers the question concerning how `inspect` enjoys its own peculiar sweet spot of strict/non-strictness in terms of matching on `refl`... cf. https://github.com/agda/agda/issues/5833

**Comment by jamesmckinna** [comment]
The other thing to ask for here is the original *working* version of the code under 2.6.3/stdlib-v?.?.?, again in as cleaned-up a state as possible.

**Comment by MatthewDaggitt** [comment]
@jamesmckinna did one of the PRs fix this or is this still an outstanding problem?

**Comment by jamesmckinna** [comment]
This is still outstanding. But I requested a look at the original code said to be working under 2.6.3, but have yet to see it. The definition of `remove` in `Data.Fin.Permutation`does typecheck, so I'm not completely clear what the exact issue is. Saying something doesn't work using `with ... in` seems moot to me since we undeprecated `inspect`. But it's possible that I have not actually understood the original problem well enough.
